### PR TITLE
Symfony Mailer Support

### PR DESCRIPTION
**Pull request by Baachi**
Hey Guys!

I created a symfony mailer reporter because swiftmailer is "official" dead I would think.
The SwiftMailer reporter is still available, so no BC breaks :)

#SymfonyHackday

### CODE REVIEW

**Comment by Baachi** [comment]
Okay it seems that I have to adjust the Travis file. The mailer component does not run on php versions lower than 7.3. :/

**Comment by kbond** [comment]
Thanks @Baachi, the travis issue is not with php < 7.3 but that `symfony/mailer` requires symfony components 4.4+.

File: Exception/MissingPackageException.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+declare(strict_types=1);
```
Comment (kbond) [inline Exception/MissingPackageException.php]: Can you remove this?

File: Helper/AbstractMailReporter.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
```
Comment (kbond) [inline Helper/AbstractMailReporter.php]: remove.

File: Helper/SymfonyMailerReporter.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
```
Comment (kbond) [inline Helper/SymfonyMailerReporter.php]: Remove.

File: Tests/Helper/SymfonyMailerReporterTest.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
```
Comment (kbond) [inline Tests/Helper/SymfonyMailerReporterTest.php]: Remove.

File: DependencyInjection/LiipMonitorExtension.php [hunk 5]
```diff
@@ -255,6 +254,26 @@ private function configureDoctrineMigrationsCheck(ContainerBuilder $container, a
...
+    private function configureMailer(ContainerBuilder $container, LoaderInterface $loader, array $config)
+    {
+        if (false === $config['mailer']['enabled']) {
+            return;
+        }
+
+        try {
+            $mailerDefinition = $container->findDefinition('mailer');
+        } catch (ServiceNotFoundException $e) {
+            throw new MissingPackageException('To enable mail reporting you have to install the "swiftmailer/swiftmailer" or "symfony/mailer".');
```
Comment (kbond) [inline DependencyInjection/LiipMonitorExtension.php]: Can we just throw a \RuntimeException here? Don't know that we need a custom exception for this.

File: Tests/DependencyInjection/LiipMonitorExtensionTest.php [hunk 6]
```diff
@@ -103,6 +112,48 @@ public function testMailer()
         $this->assertContainerBuilderHasService('liip_monitor.reporter.swift_mailer');
     }
 
+    public function testSymfonyMailer()
```
Comment (kbond) [inline Tests/DependencyInjection/LiipMonitorExtensionTest.php]: mark as skipped if `MailerInterface` does not exist.

File: composer.json [hunk 7]
```diff
@@ -42,7 +42,8 @@
         "symfony/asset": "^3.4|^4.0",
         "symfony/templating": "^3.4|^4.0",
         "phpunit/phpunit": "^7.0",
-        "symfony/finder": "^3.4|^4.0"
+        "symfony/finder": "^3.4|^4.0",
+        "symfony/mailer": "^4.3 || ^5.0"
```
Comment (kbond) [inline composer.json]: Remove this, we'll add with travis if symfony > 4.3

**Comment by kbond** [comment]
Thanks @Baachi, I made these changes.

File: DependencyInjection/LiipMonitorExtension.php [hunk 8]
```diff
@@ -255,6 +253,26 @@ private function configureDoctrineMigrationsCheck(ContainerBuilder $container, a
...
     }
 
+    private function configureMailer(ContainerBuilder $container, LoaderInterface $loader, array $config)
+    {
+        if (false === $config['mailer']['enabled']) {
+            return;
+        }
+
+        try {
+            $mailerDefinition = $container->findDefinition('mailer');
```
Comment (ixarlie) [inline DependencyInjection/LiipMonitorExtension.php]: Hi @Baachi , @kbond 

I am unable to make the bundle work when I enable the mailer configuration.

I am using Symfony 4.4 (symfony/mailer 4.4, symfony/mailgun-mailer 4.4), LiipMonitorBundle 2.11.1

Regarding the [Symfony documentation](https://symfony.com/doc/4.3/bundles/extension.html#using-the-load-method)
> In the load() method, all services and parameters related to this extension will be loaded. This method doesn't get the actual container instance, but a copy. This container only has the parameters from the actual container. After loading the services and parameters, the copy will be merged into the actual container, to ensure all services and parameters are also added to the actual container.

The container copy won't never have any mailer definition, so the exception is always raise.

Does it need a CompilerPass to load the mailer configuration instead?

Thanks!

Comment (kbond) [reply DependencyInjection/LiipMonitorExtension.php]: Good catch @ixarlie, it looks like we should just be setting the config parameters in the extension and like you suggest, have a compiler pass to create the service.

@Baachi, are you able to take a look?

Comment (ixarlie) [reply DependencyInjection/LiipMonitorExtension.php]: Hi @kbond , @Baachi  finally I've opened a new pull request addressing this issue. https://github.com/liip/LiipMonitorBundle/pull/227

**Pull request merged**